auditing/auditing.go

```diff
@@ -0,0 +1,97 @@
+package auditing
+
+import (
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	Component        string
+	URL              string
+	APIKey           string
+	IndexPrefix      string
+	RotationInterval Interval
+	Keep             int64
+	Log              *zap.SugaredLogger
+}
+
+type Interval string
+
+var (
+	HourlyInterval  Interval = "@hourly"
+	DailyInterval   Interval = "@daily"
+	MonthlyInterval Interval = "@monthly"
+)
+
+type EntryType string
+
+const (
+	EntryTypeHTTP  EntryType = "http"
+	EntryTypeGRPC  EntryType = "grpc"
+	EntryTypeEvent EntryType = "event"
+)
+
+type EntryDetail string
+
+const (
+	EntryDetailGRPCUnary  EntryDetail = "unary"
+	EntryDetailGRPCStream EntryDetail = "stream"
+)
+
+type EntryPhase string
+
+const (
+	EntryPhaseRequest  = "request"
+	EntryPhaseResponse = "response"
+	EntryPhaseSingle   = "single"
+	EntryPhaseError    = "error"
+	EntryPhaseOpened   = "opened"
+	EntryPhaseClosed   = "closed"
+)
+
+type Entry struct {
+	Id        string // filled by the auditing driver
+	Component string
+	RequestId string `json:"rqid"`
+	Type      EntryType
+	Timestamp time.Time
+
+	User   string
+	Tenant string
+
+	// For `EntryDetailHTTP` the HTTP method get, post, put, delete, ...
+	// For `EntryDetailGRPC` unary, stream
+	Detail EntryDetail
+	// e.g. Request, Response, Error, Opened, Close
+	Phase EntryPhase
+	// For `EntryDetailHTTP` /api/v1/...
+	// For `EntryDetailGRPC` /api.v1/... (the method name)
+	Path         string
+	ForwardedFor string
+	RemoteAddr   string
+
+	Body       any // JSON, string or []byte
+	StatusCode int // only for `EntryDetailHTTP`
+
+	// Internal errors
+	Error error
+}
+
+func (e *Entry) NextPhase() {
+	e.Id = ""
+	e.Timestamp = time.Now()
+	e.Body = nil
+	e.Error = nil
```

Comment:
Maybe it's a bit unexpected from the name of this function to assume that fields are cleared. Either private declaration or a different naming would be nice.

Reply:
Good point.

Alternative name: `PrepareForNextPhase` or `PrepareForResponse` (more clear, but "wrong")
Alternative API: `func (e Entry) NewEntryForNextPhase() Entry` used as `e = e.NewEntryForNextPhase()` (or with the slightly inaccurate `Response` instead of `NextPhase`).

Reply:
The alternative sounds pretty good to me. Why do we need it to be public in the first place though?

Reply:
I thought about providing a helper for those calls. But you are right. It doesn't need to be public as metal-lib is opinionated and not general purpose.